File: leetcode/0071-simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        var arr = path.split("/");
        var d = new ArrayDeque<String>();
        
        for(var p : arr) {
            if(p.isEmpty() || p.equals(".")) continue;
            else if(p.equals("..")) {if(!d.isEmpty()) d.pollLast();}
            else d.addLast(p);
        }
        
        var sb = new StringBuilder();
        while(!d.isEmpty()) {
            sb.append("/");
            sb.append(d.pollFirst());
        }
        
        return sb.isEmpty() ? "/" : sb.toString();
    }
}

File: 0071-simplify-path/0071-simplify-path.js
/**
 * @param {string} path
 * @return {string}
 */
function simplifyPath(path) {
    const arr = path.split("/");
    const stack = [];

    for (const p of arr) {
        if (p === "" || p === ".") {
            continue;
        } else if (p === "..") {
            if (stack.length > 0) {
                stack.pop();
            }
        } else {
            stack.push(p);
        }
    }

    return stack.length === 0 ? "/" : "/" + stack.join("/");
}